File: examples/android-map/src/main/kotlin/org/maplibre/nativeffi/examples/androidmap/Viewport.kt
package org.maplibre.nativeffi.examples.androidmap

import android.util.Log
import kotlin.math.ceil
import kotlin.math.max
import org.maplibre.nativeffi.render.RenderTargetExtent

internal data class Viewport(
  val logicalWidth: Int,
  val logicalHeight: Int,
  val physicalWidth: Int,
  val physicalHeight: Int,
  val scaleFactor: Double,
  val isEmpty: Boolean,
) {
  val extent: RenderTargetExtent
    get() = RenderTargetExtent(logicalWidth, logicalHeight, scaleFactor)

  fun log(label: String) {
    Log.i(
      TAG,
      "$label: logical=${logicalWidth}x$logicalHeight physical=${physicalWidth}x$physicalHeight scale=$scaleFactor empty=$isEmpty",
    )
  }

  companion object {
    private const val TAG = "MapLibreViewport"

    fun fromView(width: Int, height: Int, density: Float): Viewport {
      val physicalWidth = max(width, 0)
      val physicalHeight = max(height, 0)
      val empty = physicalWidth == 0 || physicalHeight == 0
      val scale = density.takeIf { it.isFinite() && it > 0f }?.toDouble() ?: 1.0
      return Viewport(
        logicalWidth = if (empty) 0 else max(1, ceil(physicalWidth / scale).toInt()),
        logicalHeight = if (empty) 0 else max(1, ceil(physicalHeight / scale).toInt()),
        physicalWidth = physicalWidth,
        physicalHeight = physicalHeight,
        scaleFactor = scale,
        isEmpty = empty,
      )
    }
  }
}

File: examples/browser-map/src/viewport.rs
use maplibre_native::{Error, ErrorKind, RenderTargetExtent};

pub struct Viewport {
    pub logical_width: u32,
    pub logical_height: u32,
    pub scale_factor: f64,
}

impl Viewport {
    pub fn new(logical_width: u32, logical_height: u32, scale_factor: f64) -> Result<Self, Error> {
        if logical_width == 0
            || logical_height == 0
            || !scale_factor.is_finite()
            || scale_factor <= 0.0
        {
            return Err(Error::new(
                ErrorKind::InvalidArgument,
                None,
                "invalid browser viewport",
            ));
        }
        Ok(Self {
            logical_width,
            logical_height,
            scale_factor,
        })
    }

    pub fn extent(&self) -> RenderTargetExtent {
        RenderTargetExtent::new(self.logical_width, self.logical_height, self.scale_factor)
    }

    pub fn physical_width(&self) -> u32 {
        physical_dimension(self.logical_width, self.scale_factor)
    }

    pub fn physical_height(&self) -> u32 {
        physical_dimension(self.logical_height, self.scale_factor)
    }
}

fn physical_dimension(logical: u32, scale: f64) -> u32 {
    ((logical as f64 * scale).ceil().max(1.0)) as u32
}
